Skip documentation publication when opam's `doc` field is missing

gpetiot opened the pull request:
Fix #149

NathanReb commented on bin/publish.ml:
```diff
@@ -31,6 +30,21 @@ let publish_doc ~dry_run ~yes pkg_names pkg =
+      | Ok (Some _) -> publish_doc ~dry_run ~yes pkg_names pkg
+      | Error _ | Ok None ->
+          Pkg.name pkg >>= fun name ->
+          App_log.status (fun l -> l "No doc field found for package %s" name);
```
Regarding the user logs, what would you think about:
- if there's is a doc field, we log `Publishing documentation` and go on with it
- if there's none, we log `Skipping documentation publication as there is no 'doc' field in xxx.opam`

I think it's a bit better than saying that we're going to publish the doc only to say we're not going to.
Might be nice to point to the opam file as well, to avoid confusion about where the doc field is.
I think it seems obvious to us but to new comers or to people that aren't well aware of dune-release's behaviour it might not.

NathanReb replied:
Maybe `Skipping documentation publication for package x: no 'doc field in x.opam` is a better message than what I suggested above!

What do you think?

NathanReb commented on lib/lint.ml:
```diff
@@ -88,10 +88,14 @@ let lint_res ~msgf = function
+  | Ok _ ->
+      report_status `Ok (fun l ->
+          l "opam field %a can be parsed by dune-release" pp_field "doc")
+  | Error _ ->
```
If we're going to allow missing doc fields, how about not reporting a failure (even though it's non fatal) if it's missing?

What would you think about the following:
- If there's no doc field, we mention it with a `Skipping doc field linting, no doc field for package xxx` or something similar, and succeed
- If there's one, we report whether we can parse it as you did there, in a non-fatal way so that it doesn't hurt people hosting there doc on readthedocs for instance.

NathanReb commented:
I'm not too worried about multi-opam repos for now, I think fixing the behaviour for the main use case is already worth something.

What about 2. though?

NathanReb commented:
As a side note, I think in the future the default behaviour for multi-opam should be: we only publish the doc of packages that have a doc field pointing to ghpages.

Also since we don't necessarily need that doc uri to be able to publish the doc (ie we can infer a good enough one from the dev-repo) might be worth to allow an escape hatch for people that may want to have their doc field pointing elsewhere but still publish their doc to gh-pages as well.

Anyway, I digress but we should open an issue to make sure we don't lose track of these improvement ideas! Could you open one?

gpetiot commented:
> What about 2. though?

I modified the code to make sure we fail when `publish doc` is invoked explicitly from the CLI and added some tests, let me know what you think

NathanReb commented on bin/publish.ml:
```diff
@@ -43,9 +66,9 @@ let publish_alt ~dry_run pkg kind =
-let publish ?build_dir ?opam ?delegate ?change_log ?distrib_uri ?distrib_file
-    ?publish_msg ~name ~pkg_names ~version ~tag ~keep_v ~dry_run
-    ~publish_artefacts ~yes () =
+let publish ?(from_cli = false) ?build_dir ?opam ?delegate ?change_log
```
This `from_cli` argument seems a bit convoluted to me. I think we could determine whether it was explicitly asked for with the `publish_artifact` argument, ie if it contains `Doc` then it was asked for explicitly. I think this makes things clearer.

I wonder if it wouldn't be worth changing `Pkg.publish_artifact` to `Pkg.default_publish_artifacts` and either use the cli args if they're non empty or the default ones. It would make things much clearer than having to populate the package with the args and always go through `Pkg`. I find it rather confusing.

gpetiot replied:
That's a good catch, I changed this.
